*annonex2embl*
==============

[![Build Status](https://travis-ci.com/michaelgruenstaeudl/annonex2embl.svg?branch=master)](https://travis-ci.com/michaelgruenstaeudl/annonex2embl)
[![PyPI status](https://img.shields.io/pypi/status/annonex2embl.svg)](https://pypi.python.org/pypi/annonex2embl/)
[![PyPI pyversions](https://img.shields.io/pypi/pyversions/annonex2embl.svg)](https://pypi.python.org/pypi/annonex2embl/)
[![PyPI version shields.io](https://img.shields.io/pypi/v/annonex2embl.svg)](https://pypi.python.org/pypi/annonex2embl/)
[![PyPI license](https://img.shields.io/pypi/l/annonex2embl.svg)](https://pypi.python.org/pypi/annonex2embl/)

Converts an annotated DNA multi-sequence alignment (in [NEXUS](http://wiki.christophchamp.com/index.php?title=NEXUS_file_format) format) to an EMBL flatfile for submission to [ENA](http://www.ebi.ac.uk/ena) via the [Webin-CLI submission tool](https://ena-docs.readthedocs.io/en/latest/cli_05.html).


## INSTALLATION
First, please be sure to have [Python 3 installed](https://www.python.org/downloads/) on your system. Then:

To get the most recent stable version of *annonex2embl*, run:

    pip install annonex2embl

Or, alternatively, if you want to get the latest development version of *annonex2embl*, run:

    pip install git+https://github.com/michaelgruenstaeudl/annonex2embl.git


## INPUT, OUTPUT AND PREREQUISITES
* **Input**: an annotated DNA multiple sequence alignment in NEXUS format; a comma-delimited metadata table
* **Output**: a submission-ready, multi-record EMBL flatfile

#### Requirements / Input preparation
The annotations of a NEXUS file are specified via [SETS-block](http://hydrodictyon.eeb.uconn.edu/eebedia/index.php/Phylogenetics:_NEXUS_Format), which is located beneath a DATA-block and defines sets of characters in the DNA alignment. In such a SETS-block, every gene and every exon charset must be accompanied by one CDS charset. Other charsets can be defined unaccompanied.

#### Example of a complete SETS-BLOCK
```
BEGIN SETS;
CHARSET matK_gene_forward = 929-2530;
CHARSET matK_CDS_forward = 929-2530;
CHARSET trnK_intron_forward = 1-928 2531-2813;
END;
```

#### Examples of corresponding DESCR variable
```
DESCR="tRNA-Lys (trnK) intron, partial sequence; maturase K (matK) gene, complete sequence"
```

## EXAMPLE USAGE
#### On Linux / MacOS
```
SCRPT=$PWD/scripts/annonex2embl_launcher_CLI.py
INPUT=examples/input/TestData1.nex
METAD=examples/input/Metadata.csv
OTPUT=examples/temp/TestData1.embl
DESCR='description of alignment here'  # Do not use double-quotes
EMAIL=your_email_here@yourmailserver.com
AUTHR='your name here'  # Do not use double-quotes
MNFTS=PRJEB00000
MNFTD=${DESCR//[^[:alnum:]]/_}

python3 $SCRPT -n $INPUT -c $METAD -d "$DESCR" -e $EMAIL -a "$AUTHR" -o $OTPUT --productlookup --manifeststudy $MNFTS --manifestdescr $MNFTD --compress
```

#### On Windows
```
SET SCRPT=$PWD\scripts\annonex2embl_launcher_CLI.py
SET INPUT=examples\input\TestData1.nex
SET METAD=examples\input\Metadata.csv
SET OTPUT=examples\temp\TestData1.embl
SET DESCR='description of alignment here'
SET EMAIL=your_email_here@yourmailserver.com
SET AUTHR='your name here'
SET MNFTS=PRJEB00000
SET MNFTD=a_unique_description_here

python %SCRPT% -n %INPUT% -c %METAD% -d %DESCR% -e %EMAIL% -a %AUTHR% -o %OTPUT% --productlookup --manifeststudy %MNFTS% --manifestdescr %MNFTD% --compress
```

## TO DO
* ~~Have each function of "ParsingOps.py" that interacts with a third-party server/cgi (i.e., _id_lookup, _gene_product_lookup, _taxname_lookup_ncbi, _taxname_lookup_ena) check first if that server/cgi is up and running. If the server/cgi is accessible, execute the function; if the server/cgi is inaccessible, raise an exception with message ("Server/CGI `XYZ` currently inaccessible. Please check connection to it."). For example, function "_gene_product_lookup" interacts with the "epost"-cgi (https://eutils.ncbi.nlm.nih.gov/entrez/eutils/epost.fcgi) through the BioPython-function "Entrez.epost" (https://biopython.org/DIST/docs/api/Bio.Entrez-pysrc.html#epost). Similarly, it interacts with the "esummary"-cgi (https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esummary.fcgi) through the Biopython function "Entrez.esummary" (https://biopython.org/DIST/docs/api/Bio.Entrez-pysrc.html#esummary). The accessibility of both CGIs needs to be confirmed before the function is executed.~~

* ~~Whenever a function interacts with a third-party server/cgi, please have this information printed to the screen. For example, the software should state "Communicating with server/CGI `XYZ` regarding the gene product of `foobar`.".~~

* ~~Include an optional commandline parameter with which a user can specify the name of the qualifier that is generated by function "regular_feat()" of "GenerationOps.py". Currently, that function has the qualifier name "note" hard-coded and as only possible value (see line 250 of GenerationOps.py). Upon making the qualifier name a user-specified value, please set "note" as the default value for the qualifier name. Moreover, please introduce an if-statement that states that if a user selects "product" or "gene" as qualifier name, yet the feature_type is CDS or gene (see lines 253 and 255), then (and only then) the qualifier name is changed to "note" so as not to conflict with the automatically introduced qualifier "product" and "gene" in lines 254 and 256, respectively.~~

* Furthermore, users must not specify "transl_table", "codon_start" or "estimated_length" as qualifier names, as these are automatically included in the output via our code.

* ~~Would it be possible to have the CDS definition of Taxon 4 in TestData2.embl changed from "complement(join(19..27,<11..16))" to "complement(join(<11..16,19..27))", while the automatic translation remains "MLSLL"? (I know that we chatted about it, but I don't remember the precise discussion anymore. Something about the meaning of "complement".) If it is possible, then please adjust the code accordingly. If not, then please write down an explanation sentence for me as to why that would mess up the translation.~~
Currently the output is "complement(join(<11..16,19..27))"

<!--
## TESTING
    python3 -m unittest discover -s tests -p "*_test.py"
    python3 -m unittest discover -s tests -p "*_test.py" -v  # verbose version
    pytest  # on Linux only, if python-pytest installed via pip
-->

## CHANGELOG
See [`CHANGELOG.md`](CHANGELOG.md) for a list of recent changes to the software.
